monitor.py

```diff
@@ -293,49 +309,45 @@ def __init__(self, _db, collection=None, batch_size=None):
         self.model_class = cls.MODEL_CLASS
         super(SweepMonitor, self).__init__(_db, collection=collection)
 
-    def run(self):
+    def run_once(self, *ignore):
         timestamp = self.timestamp()
         offset = timestamp.counter
         new_offset = offset
+        exception = None
 
+        # The timestamp for a SweepMonitor is purely informative --
+        # we're not trying to capture all the events that happened
+        # since a certain time -- so we're going to make sure the
+        # timestamp is set from the start of the run to the end of the
+        # last _successful_ batch.
         run_started_at = datetime.datetime.utcnow()
-        exception = None
-        achievements = None
+        timestamp.start = run_started_at
+
         while True:
-            batch_started_at = time.time()
             old_offset = offset
-            try:
-                # TODO: Change process_batch to return achievement
-                # info as well as offset, and keep track of the total.
-                new_offset = self.process_batch(offset)
-            except Exception, e:
-                self.log.error("Error during run: %s", e, exc_info=e)
-                exception = traceback.format_exc()
-                break
-            self._db.commit()
-
-            if old_offset != new_offset:
-                batch_ended_at = time.time()
-                self.log.debug(
-                    "%s monitor went from offset %s to %s in %.2f sec",
-                    self.service_name, offset, new_offset,
-                    (batch_ended_at-batch_started_at)
-                )
+            batch_started_at = datetime.datetime.utcnow()
+            new_offset = self.process_batch(offset)
+            batch_ended_at = datetime.datetime.utcnow()
+
+            self.log.debug(
+                "%s monitor went from offset %s to %s in %.2f sec",
+                self.service_name, offset, new_offset,
+                (batch_ended_at-batch_started_at).total_seconds()
+            )
 
             offset = new_offset
             if offset == 0:
                 # We completed a sweep. We're done.
-                self.cleanup()
                 break
 
-        # We're done with this run, either because we completed
-        # a sweep or because an exception was raised.
-        run_ended_at = datetime.datetime.utcnow()
-        timestamp.update(
-            start=run_started_at, finish=run_ended_at,
-            achievements=achievements, counter=new_offset,
-            exception=exception
-        )
```

Comment:
We no longer need to update the timestamp or catch an exception in this method -- we return a TimestampData with the new `counter` if we succeed, and if we raise an exception, `run` will catch it.